### PR TITLE
[#6794] std::error_code::message() reports error code name (main)

### PR DESCRIPTION
**Pull request by korydraughn**
Supersedes #6803

### CODE REVIEW

File: lib/core/include/irods/library_features.h
Comment (korydraughn) [inline lib/core/include/irods/library_features.h:30]: See TODO

File: lib/core/include/irods/replica.hpp
Comment (korydraughn) [inline lib/core/include/irods/replica.hpp:27]: See FIXME comment

File: unit_tests/src/test_system_error.cpp
Comment (korydraughn) [inline unit_tests/src/test_system_error.cpp:31]: See TODO comments

Comment (korydraughn) [reply unit_tests/src/test_system_error.cpp]: Introduced utility functions for isolating irods/errno codes.

File: lib/core/include/irods/library_features.h
Comment (korydraughn) [reply lib/core/include/irods/library_features.h:30]: This is actually fine.

Will remove the TODO comment.

Comment (korydraughn) [reply lib/core/include/irods/library_features.h]: Done.

File: unit_tests/src/test_system_error.cpp
Comment (korydraughn) [inline unit_tests/src/test_system_error.cpp:55]: How do we feel about the `get_` prefix?

Does `extract_` seem better? I feel it expresses intent better at the cost of more characters.

I originally planned to use `strip_`, but that inverts the meaning and takes more effort to understand. For example, `strip_errno(errc)` means _give me the iRODS error code without the errno_.

File: lib/core/include/irods/system_error.hpp
Comment (alanking) [inline lib/core/include/irods/system_error.hpp:56]: Can we put the word "error" somewhere in the name to indicate what this is for? Right now, it will just show up as `irods::experimental::irods_category` which is vague to me.

File: unit_tests/src/test_system_error.cpp
Comment (trel) [reply unit_tests/src/test_system_error.cpp:55]: `extract_` seems the least confusing and the most specific.

Comment (korydraughn) [reply unit_tests/src/test_system_error.cpp:55]: `extract_` it is.

File: lib/core/include/irods/system_error.hpp
Comment (alanking) [inline lib/core/include/irods/system_error.hpp:64]: 4.2.12?

File: unit_tests/src/test_system_error.cpp
Comment (alanking) [reply unit_tests/src/test_system_error.cpp:55]: I personally like `get_` more because `extract_` implies a modification to the object (by definition, "extract" means "to remove or take out"). What intent does `extract_` express that `get_` is not expressing? I'm not strongly opinionated on this, just want to understand the appeal :)

Comment (trel) [reply unit_tests/src/test_system_error.cpp:55]: 'implies a modification'... oh, does it?   if so, then `get_` is more correct / fine with me.

Comment (alanking) [reply unit_tests/src/test_system_error.cpp:55]: Well, it implies a modification to me. There may be some standard usage of "extract" of which I am currently unaware

Comment (korydraughn) [reply unit_tests/src/test_system_error.cpp:55]: I didn't know that was the meaning of extract. I'll definitely remember that for next time.

`get` can mean anything, but does work. I was trying to express that the function will return an embedded value without modification.

If there is a word that can convey that, that would be nice.

File: lib/core/include/irods/system_error.hpp
Comment (korydraughn) [reply lib/core/include/irods/system_error.hpp:56]: This is mirroring the scene used by the C++ std lib (e.g. `generic_category`, `system_category`, `future_category`, `iostream_category`). No one should be reaching for this in 99% of the cases.

Developers should prefer `make_error_code` instead of constructing `std::error_code` objects themselves. That function will call `irods_category` for you.

Does that make sense?

File: lib/core/include/irods/system_error.hpp
Comment (korydraughn) [reply lib/core/include/irods/system_error.hpp:64]: This function was part of the filesystem lib. It was always available, but usage was not advertised because it was seen as an implementation detail at the time.

With that said, do you feel it is more correct to still change it?

File: lib/core/include/irods/system_error.hpp
Comment (alanking) [reply lib/core/include/irods/system_error.hpp:56]: Oooh, I did not know that. Okay, then I'm fine with this. I don't like it, but that's just how it is for these categories!

File: lib/core/include/irods/system_error.hpp
Comment (alanking) [reply lib/core/include/irods/system_error.hpp:64]: I view this more as a removal and adding a new function (even though it is functionally identical), but I only pointed it out because I noticed that it was different. I'm fine with it either way

File: unit_tests/src/test_system_error.cpp
Comment (trel) [reply unit_tests/src/test_system_error.cpp:55]: `get` seems great.

File: lib/core/include/irods/system_error.hpp
Comment (korydraughn) [reply lib/core/include/irods/system_error.hpp:64]: After some discussion, we've landed on 4.2.12

File: unit_tests/src/test_system_error.cpp
Comment (korydraughn) [reply unit_tests/src/test_system_error.cpp:55]: Moving forward with `get`.

Comment (korydraughn) [reply unit_tests/src/test_system_error.cpp:55]: Done.

File: lib/core/include/irods/system_error.hpp
Comment (korydraughn) [reply lib/core/include/irods/system_error.hpp]: Done.

File: lib/core/include/irods/replica.hpp
Comment (korydraughn) [reply lib/core/include/irods/replica.hpp:27]: The question now is how do I address this? Hmm.

Seems my only option is to introduce another file and move the functions into it. Maybe `irods/filesystem/path_utilities.hpp`? Or maybe the detail header becomes `path_utilities.hpp`.

Comment (korydraughn) [reply lib/core/include/irods/replica.hpp:27]: The only problem with that is if someone modifies the behavior in the future and it results in a change in behavior for the fsys lib.

Now, if the replica lib is an extension of the fsys lib, then maybe we don't have to change anything.

Thoughts?

Comment (korydraughn) [reply lib/core/include/irods/replica.hpp:27]: The tests for these libraries aren't exhaustive as far as exceptions are concerned. How would we detect a behavior change before release?

The other thing is that having a detail namespace or header for a lib is signal that those types/functions/etc are private to the implementation and shouldn't be used. I feel we have to separate them to future proof ourselves.

I really don't want to duplicate work, but maybe that is okay here because the functions we're talking about aren't complex.

Comment (alanking) [reply lib/core/include/irods/replica.hpp:27]: I feel that a shared space for these types of functions would still be good. Why are these functions in the detail namespace? I don't think I understand the concern with exposing them in something like `path_utilities.hpp`.

Comment (korydraughn) [reply lib/core/include/irods/replica.hpp:27]: After some discussion, the plan is to move the functions into a new header file, `filesystem/path_utilities.hpp`.

Comment (korydraughn) [reply lib/core/include/irods/replica.hpp]: Done.

**Comment by korydraughn** [comment]
Running unit tests.

**Comment by korydraughn** [comment]
The unit tests passed, but I'm going to rerun things due to recent changes.

**Comment by korydraughn** [comment]
All unit tests pass.

Running core tests.

**Comment by korydraughn** [comment]
Fixed core test failure and all core tests passed.

File: lib/filesystem/include/irods/filesystem/path_utilities.hpp
Comment (korydraughn) [inline lib/filesystem/include/irods/filesystem/path_utilities.hpp:29]: Codacy is correct. Not only that, but the operator should be `>=` because `MAX_NAME_LEN` represents the buffer size including a null terminating byte, but `std::strlen` does not.

Comment (korydraughn) [reply lib/filesystem/include/irods/filesystem/path_utilities.hpp]: Replaced `std::strlen` with `strnlen` to account for the fixed size.

**Comment by korydraughn** [comment]
Running tests one last time.

**Comment by korydraughn** [comment]
All unit tests and core tests passed.

Will add pounds once approved.

**Comment by korydraughn** [comment]
Added directives for some of the clang-format and clang-tidy messages.

Waiting for approvals.

**Comment by trel** [comment]
looks good by inspection

**Comment by trel** [comment]
i notice 4.2.12 in this PR.   but it's for main.   in case that's news to you.

**Comment by korydraughn** [comment]
> i notice 4.2.12 in this PR. but it's for main. in case that's news to you.

That is correct/expected. This code change will be part of 4.2.12 as well.

File: lib/filesystem/include/irods/filesystem/path_utilities.hpp
Comment (alanking) [inline lib/filesystem/include/irods/filesystem/path_utilities.hpp:17]: Should this be `\since 4.2.12`, or does Doxygen pick this up?

Comment (korydraughn) [reply lib/filesystem/include/irods/filesystem/path_utilities.hpp:17]: That's a typo. Will fix

Comment (korydraughn) [reply lib/filesystem/include/irods/filesystem/path_utilities.hpp]: Done

**Comment by korydraughn** [comment]
Any last comments before I add the #'s?

**Comment by trel** [comment]
please to #

**Review by alanking — DISMISSED** [review]

**Comment by trel** [comment]
clang format is not happy

**Comment by korydraughn** [comment]
I'm ignoring remaining clang-format messages because they lead to improper formatting.
I'm ignoring the clang-tidy message because it is not necessary for that file.

**Comment by korydraughn** [comment]
Added #'s

**Review by alanking — APPROVED** [review]

**Pull request merged**
